### PR TITLE
Add a bunch of limits.

### PR DESCRIPTION
**Pull request by Kangz**
- `minUniformBufferOffsetAlignment` at 256. Required from all APIs.
- `minStorageBufferOffsetAlignment` at 256. Required for most Vulkan
  Android devices.
- `maxVertexOutputComponents` at 64. Required for early Apple GPUs
  on Metal.
- `maxFragmentInputComponents` at 60. Required for early Apple GPUs
  on Metal.
- `maxColorAttachments` at 4. Required for early Apple GPUs on Metal
  and many Android Vulkan devices.
- `maxComputeWorkgroupStorageSize` at 16352. Required for early
  Apple GPUs on Metal and many Android Vulkan devices.
- `maxComputeWorkgroupInvocations` at 256. Required for many Android
  Vulkan devices. Chose to not have per-dimension limits as they would
  be more constraining and can be worked around when translating from
  WGSL to SPIR-V.
- `maxComputePerDimensionDispatchSize` at 65535. Required for many
  Android Vulkan devices.

See #1343

***
<a href="https://pr-preview.s3.amazonaws.com/Kangz/gpuweb/pull/1863.html" title="Last updated on Jun 24, 2021, 6:15 PM UTC (0003eb1)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/1863/528f011...Kangz:0003eb1.html" title="Last updated on Jun 24, 2021, 6:15 PM UTC (0003eb1)">Diff</a>

### CODE REVIEW

**Comment by kainino0x** [comment]
#1069

**Comment by jdashg** [comment]
Ideally we can require `maxColorAttachments: 8`, and probably don't need it as a limit, since no one seems to be offering >8.

We might want `maxVertexOutputComponents` and `maxFragmentInputComponents` to match. We're unsure about merging them into one var in case adding new shader stages causes a merged name to get confusing or obsolete.
E.g., in GL, MAX_TEXTURE_IMAGE_UNITS ended up being for *frag* shaders only, vs MAX_VERTEX_TEXTURE_IMAGE_UNITS and MAX_COMBINED_TEXTURE_IMAGE_UNITS.

**Comment by kvark** [comment]
I think we should unite the in/out location limits. Quickly glancing over them on Android Vulkan shows 64 being a safe bet.

And yes, max color attachments should just be a hard limit in the spec.

The `minUniformBufferOffsetAlignment` and `minStorageBufferOffsetAlignment` limits could be a bit dangerous, but I'm hoping that people realize they are opting into non-portability if they request them lower than 256.

**Comment by jdashg** [comment]
Maybe we should put 🌶️ or 🌶️🌶️ next to the spicier limits with respect to portability, like on a restaurant menu. :)

**Comment by Kangz** [comment]
> We might want maxVertexOutputComponents and maxFragmentInputComponents to match. We're unsure about merging them into one var in case adding new shader stages causes a merged name to get confusing or obsolete.

For the record, we also discussed whether we can believe Metal's 60 limitations for fragment shaders or not given use of the position builtin would probably cost 0.

> The minUniformBufferOffsetAlignment and minStorageBufferOffsetAlignment limits could be a bit dangerous, but I'm hoping that people realize they are opting into non-portability if they request them lower than 256.

That should be the case for all limits.

File: spec/index.bs [hunk 1]
```diff
@@ -1187,6 +1187,23 @@ The default is used if a value is not explicitly specified in {{GPUDeviceDescrip
         is {{GPUBufferBindingType/"storage"}}
         or {{GPUBufferBindingType/"read-only-storage"}}.
 
+    <tr><td><dfn>minUniformBufferOffsetAlignment</dfn>
```
Comment (kvark) [inline spec/index.bs:1190]: I wonder if there is a correlation between the uniform and storage offset alignment. Any idea?
Perhaps, we could expose a single limit here instead of two.

File: spec/index.bs [hunk 2]
```diff
@@ -1187,6 +1187,23 @@ The default is used if a value is not explicitly specified in {{GPUDeviceDescrip
...
+        <td>{{GPUSize32}} <td>Lower <td>256
+    <tr class=row-continuation><td colspan=4>
+        The required alignment for {{GPUBufferBinding/offset}} for bindings with a
+        {{GPUBindGroupLayoutEntry}} |entry| for which
+        |entry|.{{GPUBindGroupLayoutEntry/buffer}}?.{{GPUBufferBindingLayout/type}}
```
Comment (kvark) [inline spec/index.bs:1195]: What about dynamic offsets? It seems to me that the limit should apply to them as well.

File: spec/index.bs [hunk 3]
```diff
@@ -1205,6 +1222,44 @@ The default is used if a value is not explicitly specified in {{GPUDeviceDescrip
...
         The maximum allowed {{GPUVertexBufferLayout/arrayStride}}
         when creating a {{GPURenderPipeline}}.
+
+    <tr><td><dfn>maxVertexOutputComponents</dfn>
+        <td>{{GPUSize32}} <td>Higher <td>64
+    <tr class=row-continuation><td colspan=4>
+        The maximum allowed number of components of output variables for a
+        vertex stage {{GPUShaderModule}} entry-point.
+
+    <tr><td><dfn>maxFragmentInputComponents</dfn>
```
Comment (kvark) [inline spec/index.bs:1232]: as discussed, let's unite them under something like `maxInterStageShaderLocations`

File: spec/index.bs [hunk 4]
```diff
@@ -1205,6 +1222,44 @@ The default is used if a value is not explicitly specified in {{GPUDeviceDescrip
...
+        <td>{{GPUSize32}} <td>Higher <td>60
+    <tr class=row-continuation><td colspan=4>
+        The maximum allowed number of components of input variables for a
+        fragment stage {{GPUShaderModule}} entry-point.
+
+    <tr><td><dfn>maxColorAttachments</dfn>
```
Comment (kvark) [inline spec/index.bs:1238]: as discussed, let's not add this one

File: spec/index.bs [hunk 5]
```diff
@@ -1205,6 +1222,44 @@ The default is used if a value is not explicitly specified in {{GPUDeviceDescrip
...
+        The maximum number of bytes used for a compute stage
+        {{GPUShaderModule}} entry-point.
+
+    <tr><td><dfn>maxComputeWorkgroupInvocations</dfn>
+        <td>{{GPUSize32}} <td>Higher <td>256
+    <tr class=row-continuation><td colspan=4>
+        The maximum value of the product of the `workgroup_size` values for a
+        compute stage {{GPUShaderModule}} entry-point.
+
+    <tr><td><dfn>maxComputePerDimensionDispatchSize</dfn>
```
Comment (kvark) [inline spec/index.bs:1258]: is there an investigation about this?

File: spec/index.bs [hunk 6]
```diff
@@ -3640,7 +3703,6 @@ A {{GPUBindGroup}} object has the following internal slots:
             **Returns:** {{GPUBindGroup}}
 
             1. Let |bindGroup| be a new valid {{GPUBindGroup}} object.
-            1. Let |limits| be |this|.{{GPUObjectBase/[[device]]}}.{{device/[[limits]]}}.{{supported limits/maxUniformBufferBindingSize}}.
```
Comment (kvark) [inline spec/index.bs:3643]: why is this removed?

File: spec/index.bs [hunk 7]
```diff
@@ -4664,7 +4738,7 @@ details.
                 - [$validating GPUProgrammableStage$]({{GPUShaderStage/FRAGMENT}},
                     |descriptor|.{{GPURenderPipelineDescriptor/fragment}},
                     |descriptor|.{{GPUPipelineDescriptorBase/layout}}) succeeds.
-                - [$validating GPUFragmentState$](|descriptor|.{{GPURenderPipelineDescriptor/fragment}}) succeeds.
+                - [$validating GPUFragmentState$](|descriptor|.{{GPURenderPipelineDescriptor/fragment}}, |device|) succeeds.
```
Comment (kvark) [inline spec/index.bs:4741]: can we pass the limits in, instead of the whole device?

File: spec/index.bs [hunk 8]
```diff
@@ -4784,10 +4862,11 @@ dictionary GPUFragmentState: GPUProgrammableStage {
 </script>
 
 <div algorithm>
-    <dfn abstract-op>validating GPUFragmentState</dfn>(|descriptor|)
+    <dfn abstract-op>validating GPUFragmentState</dfn>(|descriptor|, |device|)
         Return `true` if all of the following requirements are met:
 
-        - |descriptor|.{{GPUFragmentState/targets}}.length must be &le; 4.
+        - |descriptor|.{{GPUFragmentState/targets}}.length must be &le;
+            |device|.limits.{{supported limits/maxColorAttachments}}.
```
Comment (kvark) [inline spec/index.bs:4869]: can we actually write `|device|.limits`? I thought we have to write `|device|.{{GPUDevice/[[limits]]}}` instead

**Comment by kainino0x** [comment]
See #1865 about having limits which exist but aren't configurable.

File: spec/index.bs (hunk 1 above)
Comment (kainino0x) [reply spec/index.bs:1190]: I don't think so, based on the PR summary which implies some devices have 256 for uniform but something smaller for storage.

File: spec/index.bs (hunk 8 above)
Comment (kainino0x) [reply spec/index.bs:4869]: I think either `|device|.{{GPUDevice/limits}}` or `|device|.{{GPUDevice/[[device]]}}.{{device/[[limits]]}}`, but the former seems fine to me.

File: spec/index.bs (hunk 4 above)
Comment (kainino0x) [reply spec/index.bs:1238]: I think let's not make it configurable, but add it depending on the resolution to #1865

File: spec/index.bs [hunk 9]
```diff
@@ -1205,6 +1222,44 @@ The default is used if a value is not explicitly specified in {{GPUDeviceDescrip
...
+        Issue: Do we need to have a max per-pixel render target size?
+
+    <tr><td><dfn>maxComputeWorkgroupStorageSize</dfn>
+        <td>{{GPUSize32}} <td>Higher <td>16352
+    <tr class=row-continuation><td colspan=4>
+        The maximum number of bytes used for a compute stage
+        {{GPUShaderModule}} entry-point.
+
+    <tr><td><dfn>maxComputeWorkgroupInvocations</dfn>
+        <td>{{GPUSize32}} <td>Higher <td>256
```
Comment (kainino0x) [inline spec/index.bs:1253]: A workaround may not be needed, unless there are a significant number of devices that support 256 but don't support (256,1,1) and (1,256,1) and (1,1,256).

File: spec/index.bs (hunk 6 above)
Comment (kainino0x) [reply spec/index.bs:3643]: Looks like this line was supposed to be edited instead of removed? (to remove the maxUniformBufferBindingSize from the end)

**Review by kainino0x — COMMENTED** [review]
LGTM modulo #1865 and #1069 (dropping support for < MTLGPUFamilyApple2)

**Review by kainino0x — COMMENTED** [review]
Still LGTM % #1865 and the open comment

File: spec/index.bs [hunk 10]
```diff
@@ -1205,8 +1224,40 @@ The default is used if a value is not explicitly specified in {{GPUDeviceDescrip
...
+        fragment stage {{GPUShaderModule}} entry-point.
+
+    <tr><td><dfn>maxComputeWorkgroupStorageSize</dfn>
+        <td>{{GPUSize32}} <td>Higher <td>16352
+    <tr class=row-continuation><td colspan=4>
+        The maximum number of bytes used for a compute stage
+        {{GPUShaderModule}} entry-point.
+
+    <tr><td><dfn>maxComputeWorkgroupInvocations</dfn>
+        <td>{{GPUSize32}} <td>Higher <td>256
```
Comment (jrprice) [inline spec/index.bs:1247]: For Metal, the workgroup size limit is per pipeline, not per device. In theory, any particular compute shader may have a lower maximum workgroup size than the device's maximum due to heavy resource usage.

I don't have any direct experience with this for Metal, so I'm just curious as to whether we are confident that we'll always be able to support a workgroup size of 256 for all possible compute shaders?

Comment (Kangz) [reply spec/index.bs:1247]: We discussed this a while ago in #275 but never resolved. In general pipeline compilation can fail for many different reasons (couldn't allocate registers, or too much spilling is another one). 256 is probably safe in most cases, but to be 100% safe developers should use `createRenderPipelineAsync` to know if the compilation succeeded.

**Comment by Kangz** [comment]
> Still LGTM % #1865 and the open comment

Do we need to block on #1865? Maybe we can revisit later?

File: spec/index.bs (hunk 3 above)
Comment (kvark) [reply spec/index.bs:1234]: how is this resolved?

File: spec/index.bs (hunk 10 above)
Comment (kvark) [reply spec/index.bs:1247]: Yeah, I suppose there is no better option than to expect 256 to be supported on Metal pipelines :/

File: spec/index.bs (hunk 3 above)
Comment (kvark) [reply spec/index.bs:1234]: `maxFragmentInputComponents` is still here

File: spec/index.bs [hunk 11]
```diff
@@ -1205,8 +1224,40 @@ The default is used if a value is not explicitly specified in {{GPUDeviceDescrip
     <tr class=row-continuation><td colspan=4>
         The maximum allowed {{GPUVertexBufferLayout/arrayStride}}
         when creating a {{GPURenderPipeline}}.
+
+    <tr><td><dfn>maxInterStageShaderLocations</dfn>
+        <td>{{GPUSize32}} <td>Higher <td>64
```
Comment (kainino0x) [inline spec/index.bs:1229]: I thought this would be 60 for now until we figure out the position thing for sure?

File: spec/index.bs [hunk 12]
```diff
@@ -1205,8 +1224,40 @@ The default is used if a value is not explicitly specified in {{GPUDeviceDescrip
     <tr class=row-continuation><td colspan=4>
         The maximum allowed {{GPUVertexBufferLayout/arrayStride}}
         when creating a {{GPURenderPipeline}}.
+
+    <tr><td><dfn>maxInterStageShaderLocations</dfn>
```
Comment (kainino0x) [inline spec/index.bs:1228]: I think this should be maxInterStageShaderComponents?

File: spec/index.bs (hunk 10 above)
Comment (kainino0x) [reply spec/index.bs:1247]: In my understanding, Metal is actually strictly _less_ restrictive than other platforms - all of the APIs can fail pipeline creation arbitrarily (e.g. because they couldn't manage to allocate all of the memory needed by the shader at a particular workgroup size). Instead, Metal is _less_ restrictive: it does not require `workgroup_size` at compile time, so they let you create a pipeline, and then tell you what the maximum `workgroup_size` is afterward\*. In other words, Metal fails softly (reduces the maximum workgroup size) while other APIs fail hard (fail to create the pipeline at all).

(\* In practice, I suspect this only helps for compute shaders which do not use workgroup memory, as workgroup memory size is almost always dependent on workgroup size.)

**Review by kainino0x — COMMENTED** [review]
Agree no need to block on #1865, but have more comments

**Review by kainino0x — APPROVED** [review]

**Review by toji — APPROVED** [review]
LGTM

**Comment by toji** [comment]
Merging as discussed with Kai in the editor's meeting today.
